crates/telio-firewall/src/firewall.rs

```diff
@@ -476,29 +509,43 @@ impl StatefullFirewall {
             return false;
         }
 
-        match ip.get_next_level_protocol() {
-            IpNextHeaderProtocols::Udp => self.handle_inbound_udp(&whitelist, &peer, &ip),
-            IpNextHeaderProtocols::Tcp => self.handle_inbound_tcp(
-                |port| whitelist.is_port_whitelisted(&peer, port),
-                peer,
-                &ip,
-            ),
+        let check_port = |port| whitelist.is_port_whitelisted(&peer, port);
+
+        match proto {
+            IpNextHeaderProtocols::Udp => self.handle_inbound_udp(check_port, &peer, &ip),
+            IpNextHeaderProtocols::Tcp => self.handle_inbound_tcp(check_port, peer, &ip),
             IpNextHeaderProtocols::Icmp => self.handle_inbound_icmp(peer, &ip),
             IpNextHeaderProtocols::Icmpv6 if self.allow_ipv6 => self.handle_inbound_icmp(peer, &ip),
             _ => false,
         }
     }
 
-    fn handle_outbound_udp<'a>(&self, ip: &impl IpPacket<'a>) {
-        let key = unwrap_option_or_return!(Self::build_udp_key(ip, false));
+    fn handle_outbound_udp<'a>(&self, pubkey: PublicKey, ip: &impl IpPacket<'a>) {
+        let link = unwrap_option_or_return!(Self::build_udp_key(ip, false));
+        let key = UdpConn { link, pubkey };
+
+        let Some(headers) = ip
+            .packet()
+            .chunks_exact(UdpConnectionInfo::LAST_PKG_HEADER)
+            .next()
+            .and_then(|c| c.try_into().ok())
+        else {
+            telio_log_warn!("Failed to extract headers of UDP packet for {key:?}");
+            return;
+        };
+
         let mut udp_cache = unwrap_lock_or_return!(self.udp.lock());
         // If key already exists, dont change the value, just update timer with entry
-        if let Entry::Vacant(e) = udp_cache.entry(key) {
-            let conninfo = UdpConnectionInfo {
-                is_remote_initiated: false,
-            };
-            telio_log_trace!("Inserting new UDP conntrack entry {:?}", e.key());
-            e.insert(conninfo);
+        match udp_cache.entry(key) {
+            Entry::Vacant(e) => {
+                let conninfo = UdpConnectionInfo {
+                    is_remote_initiated: false,
+                    last_out_pkg_headers: Some(headers),
+                };
+                telio_log_trace!("Inserting new UDP conntrack entry {:?}", e.key());
+                e.insert(conninfo);
+            }
+            Entry::Occupied(mut o) => o.get_mut().last_out_pkg_headers = Some(headers),
```

Comment:
Question: how does the logic of whitelisting work..? My question is also targeted to the original code as it follows the same logic.

My confusion comes from the fact we set `is_remote_initiated` when we receive inbound UDP connection not in cache and then later we check for it and check if port is whitelisted, what is the reason for that?

As from my understanding we should do the port whitelisting on the initial request and it's presence in the cache indicates it already passed the firewall successfully.

Reply:
I'm not entirely sure. I tried not to change the whitelisting logic. However, port whitelisting is for the incoming packets only it seems

Reply:
Fair enough, this PR doesnt look like it changes that behaviour. @Jauler @packgron could you shed some light on the whitelisting?

Reply:
> My confusion comes from the fact we set is_remote_initiated when we receive inbound UDP connection not in cache and then later we check for it and check if port is whitelisted, what is the reason for that?

For non-whitelisted peers we want to allow responses for locally initiated UDP flows (or TCP connections), but drop any packet for remotely initiated ones. Therefore we remember the 5-tuple (proto, src ip, src port, dst ip, dst port) in the LRU cache, which indicates that the connection is locally initiated to later allow responses on the same tuple (just of course with swapped src and dst).

**Note:** before stale connection killing there was no reason to do connection tracking for whitelisted peers (thus it was not done), as the packets would get accepted without consulting the connection tracking state. Right now we need that in order to have enough information when injecting TCP RST's or ICMP destination unreachable upon ExitNode reconnect.

Reply:
Thanks for the explanation.

I think the logic is quite convoluted around the `is_remote_initiated` flag, the state machine is quite complicated that given several re-reads I still have trouble understanding it fully.

I think this part in `handle_inbound_udp:613` is not needed:
```
!is_port_whitelisted(key.link.local_port)
```

As there's no way for `is_remote_initiated` to become `true` but not have port whitelisted unless the whitelisting was modified mid-flight, is this possible..?

Reply:
> unless the whitelisting was modified mid-flight, is this possible..?

Yes, it is possible.

Reply:
The conclusion is that we still need to check for the port being whitelisted